[DOC] ADR sur la gestion des erreurs dans l'API et IHM. (#1730)
1024pix/pix: docs/adr/0013-manage-error-in-api-ihm.md

                  status: errorHttpStatusCode.toString(),
                  title: 'Bad request',
                  detail: 'The server could not understand the request due to invalid syntax.',
                });

**Point 3:** Coté IHM, il est difficile d'avoir une gestion uniforme de l'erreur si cet objet a plusieurs représentations différentes.

**Point 4:** Aujourd'hui, coté IHM, plusieurs cas de figure :

- On ne gère pas l'erreur, et l'écran reste figé.
 
  async _authenticateWithUpdatedPassword({ login, password }) {
     const scope = 'mon-pix';
     try {
       await this.session.authenticate('authenticator:oauth2', { login, password,  scope });
     } catch (response) {
       this.authenticationHasFailed = true;
     }
   }
 
- On affiche un message générique pour tous les types d'erreurs.
 
   async addOrganization(event) {
     event.preventDefault();
     try {
       await this.model.save();
       this.notifications.success('L’organisation a été créée avec succès.');
       this.transitionToRoute('authenticated.organizations.get', this.model.id);
     } catch (error) {
       this.notifications.error('Une erreur est survenue.');
     }
   }
   

> - Afficher un message unique à l'utilisateur final pour tous les cas d'erreur peut l'induire en error. 

Reviewer: ```suggestion
- Afficher un message unique à l'utilisateur final pour tous les cas d'erreur peut le perturber et le diriger sur une mauvaise cause de l'erreur. 
```

>  Exemple: Afficher le message 'identifiant incorrect' lors d'une indisponibilité temporaire de l'application a déjà poussé l'utilisateur a réinitialisé son mot de pass.

Reviewer: ```suggestion
 Exemple: Afficher le message 'identifiant incorrect' lors d'une indisponibilité temporaire de l'application a déjà poussé l'utilisateur a réinitialiser son mot de pass.
```

 
- On se base sur le status code http, pour différencier les erreurs.
 
   _manageErrorsApi(firstError) {
     const statusCode = _.get(firstError, 'status');
     this.errorMessage = this._showErrorMessages(statusCode);
   }
 
   _showErrorMessages(statusCode) {
     const httpStatusCodeMessages = {
       '400': ENV.APP.API_ERROR_MESSAGES.BAD_REQUEST.MESSAGE,
       '401': ENV.APP.API_ERROR_MESSAGES.LOGIN_UNAUTHORIZED.MESSAGE,
       '500': ENV.APP.API_ERROR_MESSAGES.INTERNAL_SERVER_ERROR.MESSAGE,
       '502': ENV.APP.API_ERROR_MESSAGES.INTERNAL_SERVER_ERROR.MESSAGE,
       '504': ENV.APP.API_ERROR_MESSAGES.GATEWAY_TIMEOUT.MESSAGE,
       'default': ENV.APP.API_ERROR_MESSAGES.INTERNAL_SERVER_ERROR.MESSAGE,
     };
     return this.intl.t(httpStatusCodeMessages[statusCode] || httpStatusCodeMessages['default']);
   }
  
**Point 5:** Pour un status http donné, on peut avoir des messages d'erreurs différentes à afficher selon le contexte. Le message à afficher devrait être calculé coté IHM à cause de l'internationalisation. Ce dernier peut nécessiter des informations non présente coté client et les récupérer depuis la réponse de l'api. 
   
## Décision

**Point 1:**

> Utiliser cette structure partout dans le code: Exemple pour les BadRequest

Reviewer: ```suggestion
Utiliser la structure suivante partout dans le code:
 - status: the HTTP status code applicable to this problem, expressed as a string value.
 - code: an application-specific error code, expressed as a string value.
 - title: a short, human-readable summary of the problem that SHOULD NOT change from occurrence to occurrence of the problem, except for purposes of localization.
 
Exemple pour les BadRequest :
```

> # 13. gestion-error-api-IHM

Reviewer 2: Dans les ADR existantes, les noms de fichier sont en français et le titre aussi,.
Ici, c'est du franglais :wink: 
Je propose
```suggestion
# 13. Communiquer erreurs entre IHM et API
```

Reviewer 2: Doublon avec https://github.com/1024pix/pix/pull/1730#pullrequestreview-462505260

Date: 2020-08-04

## État

Proposed

## Contexte

> 

Reviewer 2: Ajouter un peu de contexte, exemple
```suggestion
La communication entre API et IHM étant implémentée en JSONAPI, cette ADR essaye de rester proche de la [solution préconisée par le standard](https://jsonapi.org/format/1.1/#error-objects)
```

**Point 1:**

> Actuellement, l'utilisation de l'objet JsonApiError n'est pas uniforme. Exemple de retour d'erreur Bad Request JOI: on utilise les deux attributs code, et status pour envoyer le status http.

Reviewer 3: ```suggestion
Actuellement, l'utilisation de l'objet JsonApiError n'est pas uniforme. Exemple de retour d'erreur Bad Request JOI: on utilise les deux attributs code, et status pour envoyer le status HTTP.
```

> Actuellement, l'utilisation de l'objet JsonApiError n'est pas uniforme. Exemple de retour d'erreur Bad Request JOI: on utilise les deux attributs code, et status pour envoyer le status HTTP.

Reviewer 4: ```suggestion
Actuellement, l'utilisation de l'objet JsonApiError n'est pas uniforme. Exemple de retour d'erreur `Bad Request JOI` : on utilise les deux attributs code, et status pour envoyer le status HTTP.
```

> Actuellement, l'utilisation de l'objet JsonApiError n'est pas uniforme. Exemple de retour d'erreur Bad Request JOI: on utilise les deux attributs code, et status pour envoyer le status HTTP.

Reviewer 2: ```suggestion
Actuellement, l'utilisation de l'objet JsonApiError n'est pas uniforme. Exemple de retour d'erreur Bad Request JOI: on utilise les deux attributs code et status pour envoyer le statut HTTP.
```

> Actuellement, l'utilisation de l'objet JsonApiError n'est pas uniforme. Exemple de retour d'erreur Bad Request JOI: on utilise les deux attributs code, et status pour envoyer le status HTTP.

Reviewer 3: ```suggestion
Actuellement, l'utilisation de l'objet JSONAPIError n'est pas uniforme. Exemple de retour d'erreur Bad Request JOI: on utilise les deux attributs code, et status pour envoyer le status HTTP.
```

          failAction: (request, h, err) => {
            const errorHttpStatusCode = 400;
            const jsonApiError = new JSONAPIError({
              status: errorHttpStatusCode.toString(),
              title: 'Bad request',
              detail: 'The server could not understand the request due to invalid syntax.',
            });
          failAction: (request, h, err) => {
            const errorHttpStatusCode = 400;   
            const jsonApiError = new JSONAPIError({
              code: errorHttpStatusCode.toString(),
              title: 'Bad request',
              detail: 'The server could not understand the request due to invalid syntax.',

>             });

Reviewer 3: En double

Author: Dans le premier exemple, on utilise l'attribut status. Dans le deuxième l'attribut code. 
J'ai mis en italique la différence pour lever l'ambiguïté.

>             });

Reviewer 2: Sauf si qqch m'échappe, c'est du code en doublon

Author: Oui mais la structure de l'objet error est différente. 
Dans le premier exemple: on envoie  _status: errorHttpStatusCode.toString()
Dans le deuxième exemple: on envoie  _code: errorHttpStatusCode.toString(),_

Je vais reformuler car je pense c'est pas assez clair

Reviewer 3: Ligne 17 : `_status: errorHttpStatusCode.toString(),`
Ligne 24 : `_code: errorHttpStatusCode.toString(),`

Reviewer 2: My bad

> **Point 2:** Le message renvoyé par l'api ne contient pas l'origine de la violation JOI.

Reviewer 3: ```suggestion
**Point 2:** Le message renvoyé par l'API ne contient pas l'origine de la violation JOI.
```

> **Point 2:** Le message renvoyé par l'API ne contient pas l'origine de la violation JOI.

Reviewer 2: ```suggestion
**Point 2:** Si  les données reçues par l'API ne respectent pas le format attendu, le message d'erreur qui est renvoyé ne mentionne pas les données incriminées ce qui rend le diagnostic plus long. Or JOI possède cette information.
```

         const jsonApiError = new JSONAPIError({
                  status: errorHttpStatusCode.toString(),
                  title: 'Bad request',
                  detail: 'The server could not understand the request due to invalid syntax.',
              });

> **Point 3:** Coté IHM, il est difficile d'avoir une gestion uniforme de l'erreur si cet objet est renvoyé sous différentes représentations.

Reviewer 3: ```suggestion
**Point 3:** Côté IHM, il est difficile d'avoir une gestion uniforme de l'erreur si cet objet est renvoyé sous différentes représentations.
```

> **Point 3:** Côté IHM, il est difficile d'avoir une gestion uniforme de l'erreur si cet objet est renvoyé sous différentes représentations.

Reviewer 4: J'ai un peu de mal à comprendre ici. 
`une gestion uniforme de l'erreur` => c'est parce qu'on récupère l'erreur sur les fronts sous différentes représentations c'est ça ?
`cet objet` => on parle du message d'erreur renvoyé par l'API ?

Je tente une reformulation (qui m'aide à comprendre, elle n'est pas forcément plus claire pour les autres) : 
```suggestion
**Point 3:** Côté IHM, comme l'erreur est envoyée sous différentes représentations (**Point 1**) et ne contient pas le détail de l'erreur (**Point 2**) il est difficile de la traiter (on fait au cas par cas pour le moment, le traitement des erreurs n'est pas mutualisé / globalisé).
```

Author: une gestion uniforme de l'erreur => c'est parce qu'on récupère l'erreur sur les fronts sous différentes représentations c'est ça ? Effectivement.
cet objet => on parle du message d'erreur renvoyé par l'API ? Oui
Je prend ta suggestion qui me parait plus claire et détaillée, merci :) .

> **Point 4:** Aujourd'hui, coté IHM, nous avons plusieurs cas de figure :

Reviewer 3: ```suggestion
**Point 4:** Aujourd'hui, côté IHM, nous avons plusieurs cas de figure :
```

> **Point 4:** Aujourd'hui, Côté IHM, nous avons plusieurs cas de figure :

Reviewer 2: ```suggestion
**Point 4:** Actuellement, plusieurs stratégies cohabitent côté IHM lorsque l'API renvoie une erreur
```

> - Ne pas gèrer l'erreur, et l'écran reste figé.

Reviewer 4: ```suggestion
- Ne pas gérer l'erreur et l'écran reste figé.
```

 
       async _authenticateWithUpdatedPassword({ login, password }) {
            const scope = 'mon-pix';
            try {
              await this.session.authenticate('authenticator:oauth2', { login, password,  scope });
            } catch (response) {
              this.authenticationHasFailed = true;
            }

>           }

Reviewer 2: C'est super que tu aies trouvé et mentionné des exemples ! 
Tu peux aller jusqu'au bout en mettant le lien GitHub sur la version de l'objet concerné

> - Afficher un message générique pour tous les types d'erreurs.

Reviewer 2: ```suggestion
- Indiquer à l'utilisateur qu'une erreur est survenue, sans mentionner la cause ni la solution.
```

 
        async addOrganization(event) {
             event.preventDefault();
             try {
               await this.model.save();
               this.notifications.success('L’organisation a été créée avec succès.');
               this.transitionToRoute('authenticated.organizations.get', this.model.id);
             } catch (error) {
               this.notifications.error('Une erreur est survenue.');
             }
           }
   
- Afficher un message générique à l'utilisateur final pour tous les cas d'erreur peut le perturber et le diriger sur une mauvaise cause de l'erreur. 
 

>  Exemple: Le message 'identifiant incorrect' est affichée lors d'une indisponibilité temporaire de l'application a déjà poussé l'utilisateur à réinitialiser son mot de passe.

Reviewer 3: ```suggestion
 Exemple: Le message 'identifiant incorrect' est affiché lors d'une indisponibilité temporaire de l'application a déjà poussé l'utilisateur à réinitialiser son mot de passe.
```

> - Se baser sur le code http, pour différencier les erreurs.

Reviewer 3: ```suggestion
- Se baser sur le code HTTP, pour différencier les erreurs.
```

 
       _manageErrorsApi(firstError) {
            const statusCode = _.get(firstError, 'status');
            this.errorMessage = this._showErrorMessages(statusCode);
          }
        

>           _showErrorMessages(statusCode) {
>             const httpStatusCodeMessages = {
>               '400': ENV.APP.API_ERROR_MESSAGES.BAD_REQUEST.MESSAGE,
>               '401': ENV.APP.API_ERROR_MESSAGES.LOGIN_UNAUTHORIZED.MESSAGE,
>               '500': ENV.APP.API_ERROR_MESSAGES.INTERNAL_SERVER_ERROR.MESSAGE,
>               '502': ENV.APP.API_ERROR_MESSAGES.INTERNAL_SERVER_ERROR.MESSAGE,
>               '504': ENV.APP.API_ERROR_MESSAGES.GATEWAY_TIMEOUT.MESSAGE,
>               'default': ENV.APP.API_ERROR_MESSAGES.INTERNAL_SERVER_ERROR.MESSAGE,
>             };
>             return this.intl.t(httpStatusCodeMessages[statusCode] || httpStatusCodeMessages['default']);
>           }

Reviewer 3: ```suggestion
      _showErrorMessages(statusCode) {
        const httpStatusCodeMessages = {
          '400': API_ERROR_MESSAGES.BAD_REQUEST.MESSAGE,
          '401': API_ERROR_MESSAGES.LOGIN_UNAUTHORIZED.MESSAGE,
          '500': API_ERROR_MESSAGES.INTERNAL_SERVER_ERROR.MESSAGE,
          '502': API_ERROR_MESSAGES.INTERNAL_SERVER_ERROR.MESSAGE,
          '504': API_ERROR_MESSAGES.GATEWAY_TIMEOUT.MESSAGE,
          'default': API_ERROR_MESSAGES.INTERNAL_SERVER_ERROR.MESSAGE,
        };
        return this.intl.t(httpStatusCodeMessages[statusCode] || httpStatusCodeMessages['default']);
      }
```

Author: Thanks, j'ai viré tous les ENV.APP

> **Point 5:** Dans le cas de la PR https://github.com/1024pix/pix/pull/1727, nous souhaitons afficher des messages appropriés selon le scénario X. Ce qui nous emmène à avoir plusieurs messages d'erreurs pour le même status http.

Reviewer 3: ```suggestion
**Point 5:** Dans le cas de la PR https://github.com/1024pix/pix/pull/1727, nous souhaitons afficher des messages appropriés selon le scénario X. Ce qui nous amène à avoir plusieurs messages d'erreurs pour le même status HTTP.
```

> Détail: Le message à afficher est calculé coté IHM à cause de l'internationalisation. Pour ce faire, l'ihm récupère des informations non présentes coté client via l'erreur renvoyée par l'api. 

Reviewer 3: ```suggestion
Détail: Le message à afficher est calculé coté IHM à cause de l'internationalisation. Pour ce faire, l'IHM récupère des informations non présentes côté client via l'erreur renvoyée par l'API. 
```

   
## Décision

**Point 1:**
- Utiliser la structure suivante partout dans le code: 

            - status: the HTTP status code applicable to this problem, expressed as a string value.
            - code: an application-specific error code, expressed as a string value.
            - title: a short, human-readable summary of the problem that SHOULD NOT change from occurrence to occurrence of the problem, except for purposes of localization.

>             Exemple pour les BadRequest:

Reviewer 4: ```suggestion
            Exemple pour les `BadRequest`:
```

            const jsonApiError = new JSONAPIError({
              status: errorHttpStatusCode.toString(),
              title: 'Bad request',
              detail: err.details[0].message,
            });

**Point 2:**

- Renvoyer la violation JOI dans l'erreur:

> Exemple: detail: err.details[0].message au lieu de detail: 'The server could not understand the request due to invalid syntax.',

Reviewer 4: ```suggestion
Exemple: `detail: err.details[0].message` au lieu de `detail: 'The server could not understand the request due to invalid syntax.'`,
```

> Exemple: detail: err.details[0].message au lieu de  detail: 'The server could not understand the request due to invalid syntax.',

Reviewer 3: ```suggestion
Exemple: detail: err.details[0].message au lieu de detail: 'The server could not understand the request due to invalid syntax.',
```

Reviewer 3: Pour les `failActions` je serai plus pour utiliser [Boom](https://github.com/hapijs/boom) pour faire des 422.

Author: Yes, j'avais utilisé Boom dans cette PR https://github.com/1024pix/pix/pull/925/commits/466e6fc7a5d8629ff3c1ffaf4d4594eff7bcafd1 pour gérer le failAction au niveau global. 

Mais en vérifiant l'objet, il génère un statusCode au lieu de status !

Boom.badData([message], [data]) 
Returns a 422 Unprocessable Entity error where:

message - optional message.
data - optional additional error data.
Boom.badData('your data is bad and you should feel bad');
Generates the following response payload:

```
{
    "statusCode": 422,
    "error": "Unprocessable Entity",
    "message": "your data is bad and you should feel bad"
}
```

**Point 3:**

> - Centraliser la gestion des erreurs coté ihm. ( Fera l'objet d'une autre ADR )

Reviewer 3: ```suggestion
- Centraliser la gestion des erreurs côté IHM. ( Fera l'objet d'une autre ADR )
```

**Point 4:**

> - Se baser sur le code http pour afficher des messages appropriés. Exemple ci-dessous pour le login:

Reviewer 3: ```suggestion
- Se baser sur le code HTTP pour afficher des messages appropriés. Exemple ci-dessous pour le login:
```

            },
            BAD_GATEWAY: {
              CODE: '502',
              MESSAGE: 'api-error-messages.internal-server-error'
            },
            GATEWAY_TIMEOUT: {
              CODE: '504',
              MESSAGE: 'api-error-messages.internal-server-error'
            },
          }
          
            _manageErrorsApi(firstError) {
              const statusCode = _.get(firstError, 'status');
              this.errorMessage = this._showErrorMessages(statusCode);
            }
          
            _showErrorMessages(statusCode) {
              const httpStatusCodeMessages = {
                '400': API_ERROR_MESSAGES.BAD_REQUEST.MESSAGE,
                '401': API_ERROR_MESSAGES.LOGIN_UNAUTHORIZED.MESSAGE,
                '500': API_ERROR_MESSAGES.INTERNAL_SERVER_ERROR.MESSAGE,
                '502': API_ERROR_MESSAGES.INTERNAL_SERVER_ERROR.MESSAGE,
                '504': API_ERROR_MESSAGES.GATEWAY_TIMEOUT.MESSAGE,
                'default': API_ERROR_MESSAGES.INTERNAL_SERVER_ERROR.MESSAGE,
              };
              return this.intl.t(httpStatusCodeMessages[statusCode] || httpStatusCodeMessages['default']);
            }
      
**Point 5:**

> - Enrichir la structure de l'objet error JSON API en restant conforme avec la spécification json-api-error: https://jsonapi.org/format/#errors 

Reviewer 4: ```suggestion
- Enrichir la structure de l'objet error JSON API en restant conforme avec la spécification (json-api-error)[https://jsonapi.org/format/#errors] 
```

> - Enrichir la structure de l'objet error json api en restant compliant avec la spécification json-api-error: https://jsonapi.org/format/#errors 

Reviewer 3: ```suggestion
- Enrichir la structure de l'objet error JSON API en restant conforme avec la spécification json-api-error: https://jsonapi.org/format/#errors 
```

        title: a short, human-readable summary of the problem that SHOULD NOT change from occurrence to occurrence of the problem, except for purposes of localization.
        detail: a human-readable explanation specific to this occurrence of the problem. Like title, this field’s value can be localized.
        source: an object containing references to the source of the error, optionally including any of the following members:
        pointer: a JSON Pointer [RFC6901] to the associated entity in the request document [e.g. "/data" for a primary data object, or "/data/attributes/title" for a specific attribute].
        parameter: a string indicating which URI query parameter caused the error.
        meta: a meta object containing non-standard meta-information about the error.

- On avait déjà le status, title, detail. On propose de l'enrichir avec les attributs code, et l'objet meta.

    Exemple ici: https://github.com/1024pix/pix/pull/1727/commits/8b3744c31457d5fb4a4f3ae0eb8a923ea85ed4ba

    L'attribut code: va contenir un code compréhensible de l'erreur, violation de règle de gestion pour identifier facilement l'erreur.

>     L'objet meta va contenir des informations spécifiques au usecase. Exemple: afficher un message calculé coté ihm avec des informations renvoyées dans l'objet.

Reviewer 3: ```suggestion
    L'objet meta va contenir des informations spécifiques au use-case. Exemple: afficher un message calculé côté ihm avec des informations renvoyées dans l'objet.
```

                const ACCOUNT_WITH_GAR_ALREADY_EXIST_FOR_THE_SAME_ORGANIZATION = 'Vous possédez déjà un compte Pix via l\'ENT dans votre établissement scolaire.<br> Connectez-vous à ce compte sinon demandez de l\'aide à un enseignant. (Code R33)';
            
                const errorsMessagesByCodes = {
                  'R11': ACCOUNT_WITH_EMAIL_ALREADY_EXIST_FOR_ANOTHER_ORGANIZATION,
                  'R12': ACCOUNT_WITH_USERNAME_ALREADY_EXIST_FOR_ANOTHER_ORGANIZATION,
                  'R13': ACCOUNT_WITH_GAR_ALREADY_EXIST_FOR_ANOTHER_ORGANIZATION,
                  'R31': ACCOUNT_WITH_EMAIL_ALREADY_EXIST_FOR_SAME_ORGANIZATION,
                  'R32': ACCOUNT_WITH_USERNAME_ALREADY_EXIST_FOR_THE_SAME_ORGANIZATION,
                  'R33': ACCOUNT_WITH_GAR_ALREADY_EXIST_FOR_THE_SAME_ORGANIZATION,
                  'default': this.intl.t(ENV.APP.API_ERROR_MESSAGES.INTERNAL_SERVER_ERROR.MESSAGE),
                };
                return (errorsMessagesByCodes[meta.displayShortCode] || errorsMessagesByCodes['default']);
              }

## Conséquences

> - Point 1: Une seule representation de l'objet Error permettra d'uniformiser le parsing des erreurs. 

Reviewer 3: ```suggestion
- Point 1: Une seule représentation de l'objet Error permettra d'uniformiser le parsing des erreurs. 
```

> - Point 2: Renvoyer le detail de l'erreur JOI va aider le consommateur à corriger le problème et faire un re-try.

Reviewer 3: ```suggestion
- Point 2: Renvoyer le detail de l'erreur JOI va aider le consommateur à corriger le problème et faire une nouvelle tentative.
```

- Point 3: Avoir une stratégie commune et centralisée des erreurs dans les applications Pix.
- Point 4: Une meilleure UX pour l'utilisateur final.

> - Point 5: Pouvoir afficher différents messages pour le même code http en se basant sur un code fonctionnel. 

Reviewer 3: ```suggestion
- Point 5: Pouvoir afficher différents messages pour le même code HTTP en se basant sur un code fonctionnel. 
```
